### PR TITLE
Fold the dataset sampler into mlcast (mlcast.sampling)

### PR DESCRIPTION
**Pull request by franchg**
## Summary

Folds the previously separate **`mlcast-dataset-sampler`** tool into the `mlcast`
package as a new **`mlcast.sampling`** subpackage. This removes the duplication
between the two repos and gives training a single in-package path from a source
radar Zarr to a sampled training set.

The design is a clean **producer → consumer** seam:

- **Producer** (`mlcast stats`) — scans a source Zarr and writes a *stats
  parquet*: an index of candidate datacubes, each tagged with `nan_count` /
  `sum` / `mean` / `frac_wet`. The parquet schema + metadata footer is the
  contract (`mlcast.sampling.stats_spec`).
- **Consumer** (training) — `SourceDataIndexedDataset` reads that index, and a
  pluggable **`Sampler`** reshapes the candidate pool at dataset init.

## What's new

**Data-prep CLI** (folded into the existing `mlcast` argparse CLI as subcommands):
- `mlcast stats <zarr>` — cumsum sliding-window scan → zstd stats parquet (CPU
  via `bottleneck`, or `--device cuda` for the GPU backend).
- `mlcast validate-stats <parquet>` — validates a parquet against the contract;
  prints the schema, metadata, and a row preview.

**Sampler registry** (`mlcast.sampling.samplers`):
- `Sampler` ABC + `@register_sampler` / `get_sampler` (mirrors the existing
  normalization registry).
- `UniformSampler(keep_fraction)` and
  `ImportanceSampler(column, q_min, scale, mean_weight)` — per-row accept/reject,
  run **once** at init so the dataset length is fixed and reused every epoch.
- Applied per split by `SourceDataDataModule` (`train_sampler` / `eval_sampler`):
  importance on train, uniform on val/test, so validation/test stay representative.

**Training dataset:**
- `SourceDataIndexedDataset` reads the parquet `index_path` (and still accepts a
  legacy CSV index, used without a sampler).

## Notable changes
- `bottleneck` added to core dependencies; the unused `fire` dependency removed.
- `mlcast` CLI startup made lazy — `mlcast -h` ~4.2s → ~0.08s by deferring the
  training stack to the `train` path.
- Example updated: `csv_path` → `index_path`.
- README gained a "Preparing training data" section documenting the workflow.

## Testing
- `uv run pytest` → **114 passed, 1 skipped** (the skip is the CUDA-only GPU
  backend parity test).
- `ruff check` clean.
- New tests cover sampler semantics + registry, the stats `_process_chunk`
  against two independent reference oracles, GPU/CPU parity, and the parquet
  contract.

## Out of scope / follow-ups
- Decommissioning the standalone `mlcast-dataset-sampler` repo (archive, or a
  thin shim pointing here).

### CODE REVIEW

**Comment by leifdenby** [comment]
This is looking really nice! I do think we should maybe name things slightly differently here, maybe we can make the step that computes the stats be called "time-sequence stats" instead? I.e. the CLI call would be `uv run mlcast timeseq-stats` instead?

I think I have also discovered a bug when trying to apply this to the DMI dataset, but maybe I am doing something incorrect. Steps to reproduce:

1. download to sample data from the DMI dataset where we know it is raining, e.g. from 2024-12-31T00:00 is the jupyter notebook
```bash
uv run examples/scripts/download_mlcast_dataset_sample.py precipitation.dmi_10_minutes --duration PT4H --start-time 2024-12-3
```
2. run time-sequence stats collection:
```bash
uv run mlcast stats data/mlcast-source-datasets/dmi/v0.1.1/dmi_10_minutes.zarr --data-var db
```

Screenshot 2026-06-30 at 13 40 41

this reports 0 datacube candidates. Even with just looking at two successive timesteps there are no valid datacubes found:

```bash
uv run mlcast stats data/mlcast-source-datasets/dmi/v0.1.1/dmi_10_minutes.zarr --data-var dbz --time-depth 2
```
Screenshot 2026-06-30 at 13 41 26

However, the data contains more than 7dBZ rainrate values:

```python
import xarray as xr

ds_dmi = xr.open_zarr("../data/mlcast-source-datasets/dmi/v0.1.1/dmi_10_minutes.zarr")
ds_dmi

da = ds_dmi.dbz.isel(time=slice(0, 10))

da.plot.imshow(x="x", y="y", col="time", col_wrap=5)
```
Screenshot 2026-06-30 at 13 42 47

Any ideas what could be going on @franchg?

**Comment by franchg** [comment]
Thanks for reviewing this @leifdenby!
The root cause is the `--time-step-minutes` default of `5`: the DMI dataset is 10-minute data, so the time-continuity check saw a "gap" between every pair of consecutive frames and rejected every candidate window. Your exact runs work once the cadence is passed:
`uv run mlcast stats data/mlcast-source-datasets/dmi/v0.1.1/dmi_10_minutes.zarr --data-var dbz --time-step-minutes 10`
I verified against the full DMI archive: one day of data with default parameters gives 182,778 datacube candidates (and the data-kind auto-detection correctly picks reflectivity with the 7 dBZ wet threshold from the dbz attrs).

The real bug was that this failure mode was silent, a misconfigured run happily wrote an empty parquet. I've pushed [76c2724](https://github.com/mlcast-community/mlcast/pull/17/commits/76c272431ed7bb56e0f6da14941f89d1657702cb) so stats now exits non-zero with a diagnostic instead. Your original command now fails with:
```
ERROR 0 gap-free time windows: expecting a 5-min step between frames (--time-step-minutes)
but the dominant observed step is 10 min (100.0% of 144 steps). If that is the dataset's
real cadence, rerun with --time-step-minutes 10.
```
Could you re-test on with the change I made?

As a follow-up (separate change) I'd like to auto-detect the cadence from the time coordinate, so the flag becomes an override rather than something you have to know.

One thought on `timeseq-stats` rename: it names only the time axis, but these are stats over sliding (time × y × x) windows, which the tool, README, and output already call datacube candidates. So my counter-proposal would be mlcast `datacube-stats` which says exactly what the stats are over, in vocabulary the project already uses, and validate-stats still reads as its companion. Another alternative woul be to use a verb/action: `mlcast scan` (scan the archive for candidates, short and clear) or `mlcast build-index` (the training dataset then consumes the parquet as index_path).
Any preference?

**Comment by noralinn** [comment]
Here's a minor suggestion: I don't like that all probabilities in the Importance Sampler depend on the most extreme cutout. To fix that we could replace w_max() either with a fixed value or another quantity that is more stable.

The cutout with the maximum weight is either a super-rare weather situation or potentially even a bad data/ clutter case. If there is a new record case when extending the dataset the number of selected cutouts would change accordingly (e.g. twice as high w_max -> only half as many cutouts selected for training).

Not sure what the best value would be... Maybe the 99 percentile or a physically meaningful value? Or one could tell the ImportanceSampler how many selected cutouts to aim for and all probabilities are scaled accordingly.



class ImportanceSampler(Sampler):
    """Keep each candidate with probability ``w / w.max()``, where the weight
    ``w = q_min + mean_weight * (1 - exp(-s / scale))`` rises with a per-row
    statistic ``s`` (the ``column``). High-statistic datacubes are kept
    preferentially and common ones thinned out, without duplication. Needs the
    chosen ``column`` (a legacy CSV index has none).

    Parameters
    ----------
    column : str
        The stats-parquet column to weight on, e.g. ``"mean"`` (default),
        ``"sum"``, or ``"frac_wet"``.
    q_min : float
        Floor weight on every candidate, keeping some low-statistic windows.
    scale : float
        Saturation scale of ``1 - exp(-s / scale)``; set it on the order of the
        column's typical magnitude (``mean``/``frac_wet`` ~ O(1); ``sum`` large).
    mean_weight : float
        Weight given to the statistic; relative to ``q_min`` it sets how hard
        low-statistic windows are thinned versus high ones.
    """

File: src/mlcast/data/source_data_datasets.py [hunk 1]
```diff
@@ -389,13 +428,23 @@ def __init__(
             storage_options=storage_options,
         )
 
-        self.coords = pd.read_csv(csv_path).sort_values("t")
+        self.coords = _load_sampling_index(index_path).sort_values("t")
```
Comment (leifdenby) [inline src/mlcast/data/source_data_datasets.py:431]: I'm a bit confused here. I am not sure `self.coords` is the best name here since what is returned from `_load_sampling_index(..)` now also contains statistics used when constructing the full set of sampling indexes actually to be used during training.

But I also have a more general point here: The samling parquet file (same with the CSVs) were created for a specific sample shape in `time` and space (`width` and `height`), but at the same time we have `time`, `width` and `height` as argument to the pytorch Dataset class (`SourceDataIndexedDataset`) we are constructing right here. I don't think that makes sense. Those are sampling parameters, and so should either a) be provided to the sampler itself or b) be stored with in the sampling index and retrieved from their (and so not provided in the config), at minimum in the latter case we should check that the sampling index file was constructed with the same sampling parameters as the arguments provided to `SourceDataIndexedDataset`.

What do you think?

File: src/mlcast/sampling/stats_spec.py
Comment (leifdenby) [inline src/mlcast/sampling/stats_spec.py:1]: I think we should call this "sample stats spec" rather than just "stats spec" since it is specifically the statistics of the individual samples

File: src/mlcast/sampling/samplers.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+
+import numpy as np
+import pandas as pd
+
+
+class Sampler(ABC):
```
Comment (leifdenby) [inline src/mlcast/sampling/samplers.py:18]: I think it would be simpler to reason about if a `Sampler` should be initiated using the precomputed stats where those are required. That would make the sampler itself self-contained

Comment (leifdenby) [reply src/mlcast/sampling/samplers.py:18]: Also, I am not sure `Sampler` is quite the right name. A "sampler" to me sounds like something that takes some input and produces individual samples from it. What this actually does is produce a set of sampling indicies. We could change it to implement `__len__` to know how many samples it contains and it could be an iterator so that it can return individual sets of sampling indices, to produce individual samples.

File: src/mlcast/sampling/commands/stats.py
Comment (leifdenby) [inline src/mlcast/sampling/commands/stats.py:1]: I think we should rename this `build-profiled-sampling-index` instead of `stats` to make it explicit that this command both
1. produces a set of indices for candidates samples (which satisfy the nan-fraction requirement), the "building a sampling index"
2. compute the statistics for each sample given by these indices, the "profiling" of the samples

Comment (leifdenby) [reply src/mlcast/sampling/commands/stats.py:1]: I would also move this to `mlcast.data.source_data.sampling`, having it in `mlcast.sampling` doesn't make it clear that this relates to "sampling of source datasets", but could be sampling somewhere else (diffusion models use a concept of "sampling" too, so we need to avoid confusion here).

**Review by leifdenby — COMMENTED** [review]
I've had a reasonably deep read now :) Could you have a look at my comments and let me know what you think?

I've started refactoring along the lines of my suggestions already (based on the assumption that you might be ok with my suggestions), but of course happy to course correct if you feel like my suggestions don't make sense.

File: src/mlcast/data/source_data_datasets.py (hunk 1 above)
Comment (franchg) [reply src/mlcast/data/source_data_datasets.py:431]: Agreed on both points.

`self.coords` → yes, maybe `self.candidates` or similar makes more sense now given that it holds the stats too.

On width / height / time_depth: you're right, this is actually a bug. The parquet already
stores these values in its metadata, but the dataset never reads them and just uses its own
defaults (256 × 256 × 24). 

I'd go with your option (b), with one small addition:

- The dataset reads `width`, `height` and `time_depth` from the index metadata. They go away from the config.
- If the user still passes them explicitly and they differ, the explicit values win, but with a loud warning. I'd like to keep this option for experiments (e.g. a 128 crop from a 256 index).
- Legacy CSV indexes have no metadata, so there the explicit values are still needed. If missing, fall back to the old defaults with a warning.

Does it make sense to you?

File: src/mlcast/sampling/stats_spec.py
Comment (franchg) [reply src/mlcast/sampling/stats_spec.py:1]: Agreed. I suggest to use the same words everywhere, together with the command
rename below. E.g. "sampling index" for the parquet file, "sample stats" for its columns,
"sampling parameters" for the metadata.

File: src/mlcast/sampling/samplers.py (hunk 2 above)
Comment (franchg) [reply src/mlcast/sampling/samplers.py:18]: > I think it would be simpler to reason about if a `Sampler` should be initiated using the precomputed stats where those are required. That would make the sampler itself self-contained

Agreed with the goal. One thing: today the sampler only holds its
parameters and no data, which is why the config can build `ImportanceSampler()` directly.
The data comes later: each split (train / val / test) loads the index, keeps only its own
time range, and then runs the sampler on that.

If the sampler takes the stats at init, the config will need a `fdl.Partial` instead, and
loading the index + splitting it in time has to move out of the dataset (into the
datamodule, or into a small "sampling index" object that holds rows + metadata).
I think it is ok, just pointing that it requires broader changes.

> Also, I am not sure `Sampler` is quite the right name. A "sampler" to me sounds like something that takes some input and produces individual samples from it. What this actually does is produce a set of sampling indicies. We could change it to implement `__len__` to know how many samples it contains and it could be an iterator so that it can return individual sets of sampling indices, to produce individual samples.

yes, the name is confusing. It also clashes with `torch.utils.data.Sampler` and with
the `DiffusionSampler` in #16.

What the class really does is decide which candidates to keep, so I'd name it after that,
e.g. `CandidateSelector`. I'd rather put `__len__` and iteration on the object holding the
selected rows, and keep the selector itself a simple rule.

One thing to write down somewhere, since it's on purpose: we select once at init and keep
that subset for the whole training, we don't draw again every epoch.

File: src/mlcast/sampling/commands/stats.py
Comment (franchg) [reply src/mlcast/sampling/commands/stats.py:1]: > I think we should rename this `build-profiled-sampling-index` instead of `stats` to make it explicit that this command both
> 
> 1. produces a set of indices for candidates samples (which satisfy the nan-fraction requirement), the "building a sampling index"
> 2. compute the statistics for each sample given by these indices, the "profiling" of the samples

Agreed, `stats` is not a good name. Yours is a bit long, I'd go for
`build-sampling-index` and `validate-sampling-index`. But no strong opinion, either is
fine. The default output filename (`stats_...parquet`) should be renamed too.

> I would also move this to `mlcast.data.source_data.sampling`, having it in `mlcast.sampling` doesn't make it clear that this relates to "sampling of source datasets", but could be sampling somewhere else (diffusion models use a concept of "sampling" too, so we need to avoid confusion here).

Agreed!

**Comment by franchg** [comment]
> I've had a reasonably deep read now :) Could you have a look at my comments and let me know what you think?
> 
> I've started refactoring along the lines of my suggestions already (based on the assumption that you might be ok with my suggestions), but of course happy to course correct if you feel like my suggestions don't make sense.

Super! Thanks for the review! I answered each comment inline. In short: I agree with all of
them, with a few small notes. Go ahead with the refactoring on top of this PR :)

**Comment by franchg** [comment]
> Here's a minor suggestion: I don't like that all probabilities in the Importance Sampler depend on the most extreme cutout. To fix that we could replace w_max() either with a fixed value or another quantity that is more stable.
> 
> The cutout with the maximum weight is either a super-rare weather situation or potentially even a bad data/ clutter case. If there is a new record case when extending the dataset the number of selected cutouts would change accordingly (e.g. twice as high w_max -> only half as many cutouts selected for training).
> 
> Not sure what the best value would be... Maybe the 99 percentile or a physically meaningful value? Or one could tell the ImportanceSampler how many selected cutouts to aim for and all probabilities are scaled accordingly.
> 
> class ImportanceSampler(Sampler): """Keep each candidate with probability `w / w.max()`, where the weight `w = q_min + mean_weight * (1 - exp(-s / scale))` rises with a per-row statistic `s` (the `column`). High-statistic datacubes are kept preferentially and common ones thinned out, without duplication. Needs the chosen `column` (a legacy CSV index has none).
> 
> ```
> Parameters
> ----------
> column : str
>     The stats-parquet column to weight on, e.g. ``"mean"`` (default),
>     ``"sum"``, or ``"frac_wet"``.
> q_min : float
>     Floor weight on every candidate, keeping some low-statistic windows.
> scale : float
>     Saturation scale of ``1 - exp(-s / scale)``; set it on the order of the
>     column's typical magnitude (``mean``/``frac_wet`` ~ O(1); ``sum`` large).
> mean_weight : float
>     Weight given to the statistic; relative to ``q_min`` it sets how hard
>     low-statistic windows are thinned versus high ones.
> """
> ```

Good point, thanks Nora!

In practice the weight has a ceiling: `1 - exp(-s / scale)` goes to 1 for large `s`, so the
weight can never go above `q_min + mean_weight`. With the default values, any cutout with
mean above ~5 mm/h is already at that ceiling. So as long as the dataset has some heavy
rain, `w_max` is basically always the same number and a new record cutout cannot change it.

But you are right that it should not depend on the data at all. The simple fix is to divide
by the fixed ceiling `q_min + mean_weight` instead of `w.max()`. Then the keep probability
of a cutout depends only on the cutout itself, not on the rest of the dataset.

What do you think?

File: src/mlcast/data/source_data_datasets.py (hunk 1 above)
Comment (leifdenby) [reply src/mlcast/data/source_data_datasets.py:431]: perfect! Yes, completely
